civiform/civiform
# [Program Card Images] Periodically remove unused program images from cloud storage via a DurableJob

### Description

This PR adds a new `DurableJob` that will delete any unused program images from cloud storage every month. This is to help save deployments some cloud storage costs. (Though, it won't save that much money because most storage websites charge based on how many times a resource is accessed, and an old program image won't be shown on the homepage anyway.)

It does this by using `S3Client`, which is part of the AWS SDK -- see https://aws.amazon.com/sdk-for-java/. We already use the SDK for other AWS actions like sending emails:https://github.com/civiform/civiform/blob/18cc897b0f3b22be1c0f8544f68cfa47c908d833/server/app/services/cloud/aws/SimpleEmail.java#L124
so I believe this pattern is fine to use for program image deletion as well.

Breakdown of this PR:

1. In `UnusedProgramImagesCleanupJob`, it figures out the list of files to keep by polling all active and draft programs for their files and sends that list to `PublicStorageClient`.
2. In `PublicStorageClient`, add a `prunePublicFileStorage` method which takes in a list of files to keep and deletes everything else. That method delegates to `AwsS3ClientWrapper`.
3. `AwsS3ClientWrapper` sends the delete request to the real AWS `S3Client` in local servers and deployments to delete the unused images. This wrapper existed so it can be faked out in tests.



## Release notes

### Checklist

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)


### Flagging

Note on flagging: I couldn't figure out how to only create the durable job if the `PROGRAM_CARD_IMAGES` flag was on, because `SettingsManifest.getProgramCardImages` takes in a `RequestHeader` which `DurableJobModule` doesn't have. I think this is fine because if the flag is off then there just won't be any files to delete.

### Instructions for manual testing

1. Follow the directions in https://github.com/civiform/civiform/wiki/Durable-Jobs to have the `UnusedProgramImagesCleanupJob` run on local server startup.
2. Run `bin/run-dev`.
1. Enable the PROGRAM_CARD_IMAGES flag.

Try out a few different scenarios:
1. Upload a program image, publish that program, then upload a new image. Go to https://civiform-local-s3-public.s3.localhost.localstack.cloud:4566/ and verify that *both* images are still listed as resources. (We need both images because the image associated with the Active version should still be shown to applicants now, and the image associated with the new Draft version will need to be shown to applications once it's published.)
2. Publish the program with the new image. Re-check the localstack URL and verify that now only the *new* image is listed as a resource.
4. Delete the new image, but don't publish. Verify that the image is still listed as a resource.
5. Publish. Verify the image is no longer listed.
6. Add images for a few different programs. Verify that all images remain listed.

### Issue(s) this completes

Epic #5676 

Fixes #6004 (subtask)
